File: Code Generation/responses/python/claude/hard/longest-common-prefix-of-k-strings-after-removal.py
from typing import List
from collections import defaultdict

class TrieNode:
    def __init__(self):
        self.children = {}
        self.count = 0

class Trie:
    def __init__(self):
        self.root = TrieNode()
    
    def insert(self, word, delta=1):
        node = self.root
        node.count += delta
        for ch in word:
            if ch not in node.children:
                node.children[ch] = TrieNode()
            node = node.children[ch]
            node.count += delta
    
    def longest_prefix_with_k(self, k):
        node = self.root
        if node.count < k:
            return 0
        length = 0
        while True:
            next_node = None
            for ch, child in node.children.items():
                if child.count >= k:
                    next_node = child
                    break
            if next_node is None:
                break
            length += 1
            node = next_node
        return length

class Solution:
    def longestCommonPrefix(self, words: List[str], k: int) -> List[int]:
        n = len(words)
        
        trie = Trie()
        for word in words:
            trie.insert(word, 1)
        
        answer = []
        for i in range(n):
            trie.insert(words[i], -1)
            if n - 1 < k:
                answer.append(0)
            else:
                answer.append(trie.longest_prefix_with_k(k))
            trie.insert(words[i], 1)
        
        return answer

File: Code Generation/responses/cpp/claude/hard/longest-common-prefix-of-k-strings-after-removal.cpp
#include <bits/stdc++.h>
using namespace std;

struct TrieNode {
    unordered_map<char, TrieNode*> children;
    int count = 0;
};

class Trie {
public:
    int k;
    TrieNode* root;
    unordered_map<int, int> prefixLengthsCount;
    set<int, greater<int>> prefixLengths;
    Trie(int k_) : k(k_), root(new TrieNode()) {}

    void insert(const string& word) {
        TrieNode* node = root;
        for (int i = 0; i < (int)word.size(); ++i) {
            char c = word[i];
            int sz = i + 1;
            auto it = node->children.find(c);
            if (it == node->children.end()) {
                TrieNode* nn = new TrieNode();
                node->children[c] = nn;
                node = nn;
            } else {
                node = it->second;
            }
            node->count += 1;
            if (node->count >= k) {
                prefixLengthsCount[sz] += 1;
                if (prefixLengthsCount[sz] == 1) {
                    prefixLengths.insert(sz);
                }
            }
        }
    }

    void erase(const string& word) {
        TrieNode* node = root;
        for (int i = 0; i < (int)word.size(); ++i) {
            char c = word[i];
            int sz = i + 1;
            node = node->children[c];
            if (node->count == k) {
                prefixLengthsCount[sz] -= 1;
                if (prefixLengthsCount[sz] == 0) {
                    prefixLengths.erase(sz);
                }
            }
            node->count -= 1;
        }
    }

    int getLongestCommonPrefix() {
        if (prefixLengths.empty()) return 0;
        return *prefixLengths.begin();
    }
};

class Solution {
public:
    vector<int> longestCommonPrefix(vector<string> words, int k) {
        int n = words.size();
        vector<int> answer(n, 0);
        
        if (n - 1 < k) return answer;
        
        Trie trie(k);
        for (const string& w : words) trie.insert(w);
        
        for (int i = 0; i < n; i++) {
            trie.erase(words[i]);
            answer[i] = trie.getLongestCommonPrefix();
            trie.insert(words[i]);
        }
        
        return answer;
    }
};